Signal/src/ViewControllers/MediaTileViewController.swift

```diff
@@ -343,11 +417,173 @@ public class MediaTileViewController: UICollectionViewController, MediaGalleryCe
+                                                                     target: self,
+                                                                     action: #selector(didTapSelect))
+        }
+    }
+
+    @objc
+    func didTapSelect(_ sender: Any) {
+        isInBatchSelectMode = true
+
+        guard let collectionView = self.collectionView else {
+            owsFail("\(logTag) in \(#function) collectionView was unexpectedly nil")
+            return
+        }
+
+        // show toolbar
+        UIView.animate(withDuration: 0.1, delay: 0, options: .curveEaseInOut, animations: {
+            NSLayoutConstraint.deactivate([self.footerBarBottomConstraint])
+            self.footerBarBottomConstraint = self.footerBar.autoPin(toBottomLayoutGuideOf: self, withInset: 0)
+
+            self.footerBar.superview?.layoutIfNeeded()
+
+            // ensure toolbar doesn't cover bottom row.
+            collectionView.contentInset.bottom += self.kFooterBarHeight
+        }, completion: nil)
+
+        // disabled until at least one item is selected
+        self.deleteButton.isEnabled = false
+
+        // Don't allow the user to leave mid-selection, so they realized they have
+        // to cancel (lose) their selection if they leave.
+        self.navigationItem.hidesBackButton = true
+    }
+
+    @objc
+    func didCancelSelect(_ sender: Any) {
+        isInBatchSelectMode = false
+
+        guard let collectionView = self.collectionView else {
+            owsFail("\(logTag) in \(#function) collectionView was unexpectedly nil")
+            return
+        }
+
+        // hide toolbar
+        UIView.animate(withDuration: 0.1, delay: 0, options: .curveEaseInOut, animations: {
+            NSLayoutConstraint.deactivate([self.footerBarBottomConstraint])
+            self.footerBarBottomConstraint = self.footerBar.autoPinEdge(toSuperviewEdge: .bottom, withInset: -self.kFooterBarHeight)
+            self.footerBar.superview?.layoutIfNeeded()
+
+            // undo "ensure toolbar doesn't cover bottom row."
+            collectionView.contentInset.bottom -= self.kFooterBarHeight
+        }, completion: nil)
+
+        self.navigationItem.hidesBackButton = false
+
+        // deselect any selected
+        collectionView.indexPathsForSelectedItems?.forEach { collectionView.deselectItem(at: $0, animated: false)}
+    }
+
+    @objc
+    func didPressDelete(_ sender: Any) {
+        Logger.debug("\(self.logTag) in \(#function)")
+
+        guard let collectionView = self.collectionView else {
+            owsFail("\(logTag) in \(#function) collectionView was unexpectedly nil")
+            return
+        }
+
+        guard let indexPaths = collectionView.indexPathsForSelectedItems else {
+            owsFail("\(logTag) in \(#function) indexPaths was unexpectedly nil")
+            return
+        }
+
+        let items: [MediaGalleryItem] = indexPaths.flatMap { return self.galleryItem(at: $0) }
```

Comment:
I thought we found that `flatMap` wasn't valid on XCode 9.2?

Reply:
You're thinking of `compactMap` which replaces `flatMap` in XCode 9.3.